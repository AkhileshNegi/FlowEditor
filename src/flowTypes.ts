import {
    IsOptionUniqueHandler,
    IsValidNewOptionHandler,
    NewOptionCreatorHandler,
    PromptTextCreatorHandler
} from 'react-select';
import { Operators } from './config/operatorConfigs';
import AssetService from './services/AssetService';
import { Types } from './config/typeConfigs';

export interface Languages {
    [iso: string]: string;
}

export interface Endpoints {
    fields: string;
    groups: string;
    contacts: string;
    flows: string;
    activity: string;
    labels: string;
    simulateStart: string;
    simulateResume: string;
}

export interface FlowEditorConfig {
    assetService?: AssetService;
    localStorage: boolean;
    languages: { [iso: string]: string };
    endpoints: Endpoints;
    flow: string;
    path?: string;
}

export interface LocalizationMap {
    [lang: string]: {
        [uuid: string]: any;
    };
}

export interface FlowDefinition {
    localization: LocalizationMap;
    language: string;
    name: string;
    nodes: FlowNode[];
    uuid: string;
    _ui: UIMetaData;
}

export interface FlowNode {
    uuid: string;
    actions: Action[];
    exits: Exit[];
    router?: Router;
    wait?: Wait;
}

export interface Exit {
    uuid: string;
    name?: string;
    destination_node_uuid?: string;
}

export enum RouterTypes {
    switch = 'switch'
}

export interface Router {
    type: RouterTypes;
    result_name?: string;
}

export interface Case {
    uuid: string;
    type: Operators;
    exit_uuid: string;
    arguments?: string[];
}

export interface SwitchRouter extends Router {
    cases: Case[];
    operand: string;
    default_exit_uuid: string;
}

export enum WaitTypes {
    exp = 'exp',
    group = 'group',
    msg = 'msg',
    flow = 'flow'
}

export interface Wait {
    type: WaitTypes;
    timeout?: number;
}

export interface Group {
    uuid: string;
    name: string;
}

export enum Methods {
    GET = 'GET',
    POST = 'POST',
    PUT = 'PUT'
}

export interface ChangeGroups extends Action {
    groups: Group[];
}

export interface Field {
    key: string;
    name: string;
}

export interface Label {
    uuid: string;
    name: string;
}

export interface Action {
    type: Types;
    uuid: string;
}

export interface SetContactField extends Action {
    field: Field;
    value: string;
}

export interface SetContactName extends Action {
    type: Types.set_contact_name;
    name: string;
}

export type SetContactProperty = SetContactName;

export type SetContactAttribute = SetContactField | SetContactProperty;

// tslint:disable-next-line:no-empty-interface
export interface Missing extends Action {}

export interface SendMsg extends Action {
    text: string;
    all_urns?: boolean;
}

export interface AddLabels extends Action {
    labels: Label[];
}

export interface SetPreferredChannel extends Action {
    language: string;
}

export interface SendEmail extends Action {
    subject: string;
    body: string;
    addresses: string[];
}

export interface SetRunResult extends Action {
    name: string;
    value: string;
    category?: string;
}

export interface Headers {
    [name: string]: string;
}

export interface CallWebhook extends Action {
    url: string;
    method: Methods;
    body?: string;
    headers?: Headers;
}

export interface StartFlow extends Action {
    flow: {
        name: string;
        uuid: string;
    };
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    languages: Array<{ [iso: string]: string }>;
}

export interface FlowPosition {
    left: number;
    top: number;
    right?: number;
    bottom?: number;
}

export interface Dimensions {
    width: number;
    height: number;
}

export enum UINodeTypes {
    split = 'split',
    subflow = 'subflow',
    webhook = 'webhook'
}

export interface UINode {
    position: FlowPosition;
    // ui type, used for split by expression, contact field, etc
    type?: UINodeTypes | Types;
}

export interface StickyNote {
    position: FlowPosition;
    title: string;
    body: string;
    color?: string;
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    stickies: { [key: string]: StickyNote };
}

export type AnyAction =
    | Action
    | ChangeGroups
    | SetContactField
    | SetContactProperty
    | SetRunResult
    | SendMsg
    | SetPreferredChannel
    | SendEmail
    | CallWebhook
    | StartFlow;

export enum ContactProperties {
    UUID = 'UUID',
    'Created By' = 'Created By',
    'Modified By' = 'Modified By',
    Org = 'Org',
    Name = 'Name',
    Language = 'Language',
    Timezone = 'Timezone',
    Email = 'Email',
    Mailto = 'Mailto',
    Phone = 'Phone',
    Groups = 'Groups',
    Facebook = 'Facebook',
    Telegram = 'Telegram'
}

export enum ResultType {
    flow = 'flow',
    field = 'field',
    group = 'group',
    label = 'label'
}

export enum ValueType {
    text = 'text',
    numeric = 'numeric',
    datetime = 'datetime',
    state = 'state',
    district = 'district',
    ward = 'ward'
}

export interface CreateOptions {
    promptTextCreator?: PromptTextCreatorHandler;
    newOptionCreator?: NewOptionCreatorHandler;
    isValidNewOption?: IsValidNewOptionHandler;
    isOptionUnique?: IsOptionUniqueHandler;
    createNewOption?: NewOptionCreatorHandler;
    createPrompt?: string;
}

export enum StartFlowArgs {
    Complete = 'C',
    Expired = 'E'
}

export enum StartFlowExitNames {
    Complete = 'Complete',
    Expired = 'Expired'
}

export enum WebhookExitNames {
    Success = 'Success',
    Failure = 'Failed',
    Unreachable = 'Unreachable'
}
