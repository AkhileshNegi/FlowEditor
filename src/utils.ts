const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {BoolMap} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase(), SNAKED_CHARS, '_', 'all');
}
