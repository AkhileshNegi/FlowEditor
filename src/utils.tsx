/** Turns a string array into a bool map */
export function toBoolMap(array: string[]): {[item: string]: boolean} {
    var map: {[item: string]: boolean} = {};
    for (let item of array) {
        map[item] = true;
    }
    return map;
}