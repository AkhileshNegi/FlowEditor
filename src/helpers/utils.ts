import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';

const request = require('sync-request');

const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase().trim(), SNAKED_CHARS, '_', 'all');
}

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export function getDisplayName(HOCName: string, Component: ComponentClass | SFC): string {
    const ComponentDisplayName = Component.displayName || Component.name || 'Component';
    return `${HOCName}(${ComponentDisplayName})`;
}

function getFlow(name: string): FlowDefinition {
    const definition = request('GET', 'base/test_flows/' + name + '.json').getBody();
    return JSON.parse(definition) as FlowDefinition;
}

export function getFavorites(): FlowDefinition {
    return getFlow('favorites');
}

export function getTest(): FlowDefinition {
    return getFlow('a4f64f1b-85bc-477e-b706-de313a022979');
}

export function dump(object: any) {
    console.log(JSON.stringify(object, null, 1));
}

export const DATA_SPEC_ATTRIBUTE_NAME: string = 'data-spec';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} matching DOM components
 */
export function getSpecWrapper(
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string
): ShallowWrapper | ReactWrapper {
    return componentWrapper.find(`[${DATA_SPEC_ATTRIBUTE_NAME}="${specName}"]`);
}

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export function validUUID(uuid: string): boolean {
    const regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
    return regex.test(uuid);
}
