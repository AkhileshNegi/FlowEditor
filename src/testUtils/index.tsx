// TODO: Remove use of Function
// tslint:disable:ban-types
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import mutate, { Query } from 'immutability-helper';
import { object } from 'prop-types';
import * as React from 'react';
import * as config from '../../__test__/config';
import { ConfigProviderContext, endpointsPT, flowPT, languagesPT } from '../config';
import { Operators } from '../config/operatorConfigs';
import { Types } from '../config/typeConfigs';
import {
    AnyAction,
    CallWebhook,
    Case,
    ChangeGroups,
    ContactProperties,
    Exit,
    FlowDefinition,
    FlowEditorConfig,
    FlowNode,
    Group,
    Methods,
    Router,
    SendEmail,
    SendMsg,
    SetContactProperty,
    StartFlow,
    StartFlowArgs,
    StartFlowExitNames,
    SwitchRouter,
    Wait,
    WaitTypes,
    Field
} from '../flowTypes';
import { AppState, createStore, initialState, SearchResult } from '../store';
import { capitalize, getBaseLanguage, merge, set } from '../utils';
import { SetContactField, SetRunResult } from '../flowTypes';

const { results: groupsResults } = require('../../assets/groups.json');

export interface Resp {
    results: Array<{ [key: string]: any }>;
    total?: number;
}

export interface QueryString {
    [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
    store: object,
    endpoints: endpointsPT,
    languages: languagesPT,
    flow: flowPT
};

export const baseState: AppState = {
    ...initialState,
    ...{
        flowContext: {
            ...initialState.flowContext,
            definition: require('../../__test__/flows/colors.json') as FlowDefinition
        }
    },
    ...{
        flowEditor: {
            ...initialState.flowEditor,
            editorUI: {
                ...initialState.flowEditor.editorUI,
                language: getBaseLanguage((config as FlowEditorConfig).languages)
            }
        }
    }
};

export const configProviderContext: ConfigProviderContext = {
    endpoints: (config as FlowEditorConfig).endpoints,
    languages: (config as FlowEditorConfig).languages,
    flow: (config as FlowEditorConfig).flow
};

export const setMock = (implementation?: (...args: any[]) => any): Query<jest.Mock> =>
    set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
    Component: React.ComponentClass | React.SFC,
    baseProps: P = {} as any,
    baseDuxState: AppState | Partial<AppState> = baseState,
    baseContext: ConfigProviderContext = configProviderContext
) => (
    shallowRender: boolean = true,
    propOverrides: Query<P | Partial<P>> = {},
    duxStateOverrides: Query<AppState | Partial<AppState>> = {},
    contextOverrides: Query<ConfigProviderContext | Partial<ConfigProviderContext>> = {},
    childContextTypeOverrides: { [key: string]: Function } = {}
) => {
    const props = mutate(baseProps, propOverrides);
    const store = createStore(mutate(baseDuxState, duxStateOverrides) as AppState);
    let context = mutate(baseContext, merge({ store }));

    if (Object.keys(duxStateOverrides).length > 0) {
        context = mutate(context, duxStateOverrides);
    }

    if (Object.keys(contextOverrides).length > 0) {
        context = mutate(context, contextOverrides);
    }

    const childContextTypes: { [contextProp: string]: Function } = mutate(
        contextTypes,
        childContextTypeOverrides
    );
    // tslint:disable-next-line:ban-types
    const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
        <Component {...props} />,
        {
            context,
            childContextTypes
        }
    );

    return {
        wrapper,
        props,
        context,
        instance: wrapper.instance()
    };
};

export const composeSpy = (obj: Object | React.ComponentClass) => (instanceMethod: string) =>
    jest.spyOn((obj as React.ComponentClass).prototype || obj, instanceMethod as any);

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () => new Promise(resolve => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance[]) =>
    spies.forEach(spy => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
    specName: string,
    attributeName: string = 'data-spec'
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
    query: Query<AppState | Partial<AppState>> = {},
    duxState = baseState
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
    Component: React.ComponentClass | React.SFC,
    baseProps: P = {} as any,
    baseDuxState: AppState | Partial<AppState> = baseState,
    baseContext: ConfigProviderContext = configProviderContext
) => ({
    setup: composeSetup<P>(Component, baseProps, baseDuxState, baseContext),
    spyOn: composeSpy(Component)
});

/**
 * Create a select control option
 */
export const genSelectOption = ({ label }: { label: string }) => ({
    label: capitalize(label.trim()),
    labelKey: 'name',
    valueKey: 'id'
});

export const genSendMsgAction = ({
    uuid = 'send_msg-0',
    text = 'Hey!',
    all_urns = false
}: {
    uuid?: string;
    text?: string;
    // tslint:disable-next-line:variable-name
    all_urns?: boolean;
} = {}): SendMsg => ({
    type: Types.send_msg,
    uuid,
    text,
    all_urns
});

export const genSendEmailAction = ({
    uuid = 'send_email-0',
    subject = 'New Sign Up',
    body = '@run.results.name just signed up.',
    emails = ['jane@exaumple.com']
}: {
    uuid?: string;
    subject?: string;
    body?: string;
    emails?: string[];
} = {}): SendEmail => ({
    uuid,
    type: Types.send_email,
    subject,
    body,
    emails
});

export const genCallWebhookAction = ({
    uuid = 'call_webhook-0',
    url = 'https://www.example.com',
    method = Methods.GET
}: {
    uuid?: string;
    url?: string;
    method?: Methods;
} = {}): CallWebhook => ({
    uuid,
    type: Types.call_webhook,
    url,
    method
});

export const genAddGroupsAction = ({
    uuid = 'add_contact_groups-0',
    groups = groupsResults
}: { uuid?: string; groups?: Group[] } = {}): ChangeGroups => ({
    uuid,
    type: Types.add_contact_groups,
    groups
});

export const genStartFlowAction = ({
    uuid = 'start_flow-0',
    flow_name = 'Colors',
    flow_uuid = 'colors-0'
}: {
    uuid?: string;
    // tslint:disable-next-line:variable-name
    flow_name?: string;
    // tslint:disable-next-line:variable-name
    flow_uuid?: string;
} = {}): StartFlow => ({
    type: Types.start_flow,
    uuid: flow_uuid,
    flow_name: capitalize(flow_name.trim()),
    flow_uuid
});

export const genSetContactPropertyAction = ({
    uuid = 'set_contact_property-0',
    property = ContactProperties.Email,
    value = 'jane@example.com'
}: {
    uuid?: string;
    property?: string;
    value?: string;
} = {}): SetContactProperty => ({
    uuid,
    property,
    value,
    type: Types.set_contact_property
});

export const genSetContactFieldAction = ({
    uuid = 'set_contact_field-0',
    field = {
        key: 'age',
        name: 'Age'
    },
    value = '25'
}: { uuid?: string; field?: Field; value?: string } = {}): SetContactField => ({
    uuid,
    field,
    value,
    type: Types.set_contact_field
});

export const genSetRunResultAction = ({
    uuid = 'set_run_result-0',
    result_name = 'Name',
    value = 'Grace',
    category = ''
}: {
    uuid?: string;
    result_name?: string;
    value?: string;
    category?: string;
} = {}): SetRunResult => ({
    uuid,
    result_name,
    value,
    category,
    type: Types.set_run_result
});

// tslint:disable-next-line:variable-name
export const genCase = ({
    uuid,
    type,
    exit_uuid,
    args = []
}: {
    uuid: string;
    type: Operators;
    exit_uuid: string;
    args?: string[];
}) => ({
    uuid,
    type,
    exit_uuid,
    arguments: args
});

export const genExit = ({
    uuid = 'exit-0',
    name = null,
    destination_node_uuid = 'node-1'
}: {
    uuid?: string;
    name?: string;
    // tslint:disable-next-line:variable-name
    destination_node_uuid?: string;
} = {}) => ({
    uuid,
    name,
    destination_node_uuid
});

// tslint:disable-next-line:variable-name
export const genWait = (type: WaitTypes, flow_uuid?: string) => ({
    type,
    ...(flow_uuid ? { flow_uuid } : {})
});

// tslint:disable-next-line:variable-name
export const genRouter = (result_name?: string): Router => ({
    type: 'switch',
    ...(result_name ? { result_name } : {})
});

export const genSwitchRouter = (
    cases: Case[],
    operand: string,
    // tslint:disable-next-line:variable-name
    default_exit_uuid: string
) => ({
    ...genRouter(),
    cases,
    operand,
    default_exit_uuid
});

export const genFlowNode = ({
    actions,
    exits,
    uuid = 'node-0',
    router = null,
    wait = null
}: {
    actions: AnyAction[];
    exits: Exit[];
    uuid?: string;
    router?: Router | SwitchRouter;
    wait?: Wait;
}): FlowNode => ({
    actions,
    exits,
    uuid,
    ...(router ? { router } : ({} as any)),
    ...(wait ? { wait } : ({} as any))
});

export const genStartFlowNode = (
    startFlowAction: StartFlow,
    uuid: string = 'start_flow_node-0',
    // tslint:disable-next-line:variable-name
    flow_uuid?: string,
    exitUUIDs: string[] = ['exit1', 'exit2']
): FlowNode =>
    genFlowNode({
        actions: [startFlowAction],
        exits: [
            genExit({
                uuid: exitUUIDs[0],
                name: StartFlowExitNames.Complete,
                destination_node_uuid: 'destination-completed'
            }),
            genExit({
                uuid: exitUUIDs[1],
                name: StartFlowExitNames.Expired,
                destination_node_uuid: 'destination-expired'
            })
        ],
        uuid,
        router: genSwitchRouter(
            [
                genCase({
                    uuid: 'start_flow_case-0',
                    type: Operators.has_run_status,
                    exit_uuid: exitUUIDs[0],
                    args: [StartFlowArgs.Complete]
                }),
                genCase({
                    uuid: 'start_flow_case-1',
                    type: Operators.has_run_status,
                    exit_uuid: exitUUIDs[1],
                    args: [StartFlowArgs.Expired]
                })
            ],
            '@child',
            null
        ),
        wait: genWait(WaitTypes.flow, flow_uuid)
    });

export const genGroupsRouterNode = (
    groups: Group[] = groupsResults,
    uuid: string = 'split_by_groups-0'
): FlowNode =>
    genFlowNode({
        actions: [],
        exits: groups.map((group, idx) =>
            genExit({ uuid: group.uuid, name: group.name, destination_node_uuid: `node-${idx}` })
        ),
        uuid,
        router: genSwitchRouter(
            groups.map((group, idx) =>
                genCase({
                    uuid: `split_by_group-${idx}`,
                    type: Operators.has_group,
                    exit_uuid: group.uuid,
                    args: [group.uuid]
                })
            ),
            '@contact.groups',
            null
        ),
        wait: genWait(WaitTypes.group)
    });

export const getGroupOptions = (groups: Group[] = groupsResults): SearchResult[] =>
    groups.map(({ name, uuid }) => ({
        name,
        id: uuid
    }));

export const getGroups = (sliceAt: number, groups: Group[] = groupsResults) =>
    groups
        .map(({ name, uuid }) => ({
            name,
            id: uuid
        }))
        .slice(sliceAt);
