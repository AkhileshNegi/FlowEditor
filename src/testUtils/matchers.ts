import { RenderNode } from '../store/flowContext';
import { dump } from '../utils';
import { Exit } from '../flowTypes';
import { Store } from 'redux';
import { AppState } from '../store';

interface MatchResult {
    message: () => string;
    pass: boolean;
}

function toHaveInboundConnections<T>(this: jest.MatcherUtils, received: RenderNode): MatchResult {
    if (Object.keys(received.inboundConnections).length > 1) {
        return {
            message: () => `Node ${received.node.uuid} has an inbound connection`,
            pass: true
        };
    }

    return {
        message: () => `Node${received.node.uuid} has no inbound connections`,
        pass: false
    };
}

function toHaveInboundFrom<T>(
    this: jest.MatcherUtils,
    received: RenderNode,
    expected: Exit
): MatchResult {
    for (const exitUUID of Object.keys(received.inboundConnections)) {
        if (exitUUID === expected.uuid) {
            return {
                message: () =>
                    `Node ${received.node.uuid} incorrectly has inbound from ${expected.uuid}`,
                pass: true
            };
        }
    }

    return {
        message: () =>
            `Node ${received.node.uuid} doesn't have inbound connection from ${expected.uuid}`,
        pass: false
    };
}

function toPointTo<T>(this: jest.MatcherUtils, received: Exit, expected: RenderNode): MatchResult {
    if (received.destination_node_uuid === expected.node.uuid) {
        return {
            message: () => `${received.uuid} incorrectly points to ${expected.node.uuid}`,
            pass: true
        };
    }

    return {
        message: () => `Expected ${received.uuid} to point to ${expected.node.uuid}`,
        pass: false
    };
}

function toHaveExitWithDestination<T>(this: jest.MatcherUtils, received: RenderNode): MatchResult {
    for (const exit of received.node.exits) {
        if (exit.destination_node_uuid !== null) {
            return {
                message: () => `Exit ${exit.uuid} in node ${received.node.uuid} has a destination`,
                pass: true
            };
        }
    }

    return {
        message: () => `All exits for ${received.node.uuid} are disconnected`,
        pass: false
    };
}

function toHaveExitThatPointsTo<T>(
    this: jest.MatcherUtils,
    received: RenderNode,
    expected: RenderNode
): MatchResult {
    for (const exit of received.node.exits) {
        if (exit.destination_node_uuid === expected.node.uuid) {
            return {
                message: () =>
                    `${exit.uuid} incorrectly has exit that points to ${expected.node.uuid}`,
                pass: true
            };
        }
    }

    return {
        message: () =>
            `Expected ${received.node.uuid} to have an exit that points to ${expected.node.uuid}`,
        pass: false
    };
}

function toHavePayload<T>(
    this: jest.MatcherUtils,
    store: any,
    actionType: string,
    expectedPayload: any
): MatchResult {
    const payload = JSON.stringify(expectedPayload);
    for (const actionTaken of store.getActions()) {
        if (actionTaken.type === actionType) {
            if (JSON.stringify(actionTaken.payload) === payload) {
                return {
                    message: () =>
                        `Result contained action type ${actionType} with payload ${payload}`,
                    pass: true
                };
            }
        }
    }

    return {
        message: () => `Could not find action type ${actionType} with payload ${payload}`,
        pass: false
    };
}

function toHaveReduxAction<T>(
    this: jest.MatcherUtils,
    store: any,
    actionType: string
): MatchResult {
    for (const actionTaken of store.getActions()) {
        if (actionTaken.type === actionType) {
            return {
                message: () => `Result contained action type ${actionType}`,
                pass: true
            };
        }
    }

    return {
        message: () => `Could not find action type ${actionType}`,
        pass: false
    };
}

expect.extend({
    toPointTo,
    toHaveExitThatPointsTo,
    toHaveInboundFrom,
    toHaveExitWithDestination,
    toHaveInboundConnections,
    toHavePayload,
    toHaveReduxAction
});
