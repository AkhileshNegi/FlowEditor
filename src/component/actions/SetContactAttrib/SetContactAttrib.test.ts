import { SetContactField, SetContactProperty } from '../../../flowTypes';
import { composeComponentTestUtils } from '../../../testUtils';
import {
    createSetContactFieldAction,
    createSetContactPropertyAction
} from '../../../testUtils/assetCreators';
import { setEmpty, titleCase } from '../../../utils';
import SetContactAttribComp, { getFieldNameMarkup } from './SetContactAttrib';

const setContactProperty = createSetContactPropertyAction();
const setContactField = createSetContactFieldAction();

const { setup } = composeComponentTestUtils<SetContactProperty | SetContactField>(
    SetContactAttribComp,
    setContactProperty as SetContactProperty
);

describe(SetContactAttribComp.name, () => {
    describe('render', () => {
        describe('helpers', () => {
            describe('getFieldNameMarkup', () => {
                it('should return emphasized property', () => {
                    expect(
                        getFieldNameMarkup(setContactProperty as SetContactProperty)
                    ).toMatchSnapshot();
                });

                it('should return emphasized field name', () => {
                    expect(
                        getFieldNameMarkup(setContactField as SetContactField)
                    ).toMatchSnapshot();
                });
            });
        });

        it("should render with 'update...' div when value prop passed", () => {
            const { wrapper, props } = setup();

            expect(wrapper.text()).toBe(
                `Update ${titleCase((props as SetContactProperty).property)} to ${props.value}`
            );
            expect(wrapper).toMatchSnapshot();
        });

        it("should render with 'clear...' div when value prop isn't passed", () => {
            const { wrapper, props } = setup(true, {
                value: setEmpty()
            });

            expect(wrapper.text()).toBe(
                `Clear value for ${titleCase((props as SetContactProperty).property)}`
            );
            expect(wrapper).toMatchSnapshot();
        });
    });
});
