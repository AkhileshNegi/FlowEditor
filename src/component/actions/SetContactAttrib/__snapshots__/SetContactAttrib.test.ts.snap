// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetContactAttribComp render helpers getFieldNameMarkup should return emphasized field name 1`] = `
<span
  className="emph"
>
  Age
</span>
`;

exports[`SetContactAttribComp render helpers getFieldNameMarkup should return emphasized property 1`] = `
<span
  className="emph"
>
  Email
</span>
`;

exports[`SetContactAttribComp render should render with 'clear...' div when value prop isn't passed 1`] = `
<div>
  Clear value for 
  <span
    className="emph"
  >
    Email
  </span>
</div>
`;

exports[`SetContactAttribComp render should render with 'update...' div when value prop passed 1`] = `
<div>
  Update 
  <span
    className="emph"
  >
    Email
  </span>
   to 
  <span
    className="emph"
  >
    jane@example.com
  </span>
</div>
`;
